### ISSUE TITLE
[turborepo]: Incorrect Task Caching Behavior

### ISSUE THREAD

**Issue by rubber-duck-software** [vercel/turbo#2855]
### What version of Turborepo are you using?

1.6.3

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Describe the Bug

Turborepo should differentiate between task dependencies and code dependencies. 

The `dependsOn` option is `The list of tasks this task depends on.` It seems natural that this option would be used to define task dependencies. However, this option is also used to define code level dependencies in a way which is both confusing and prevents even some simple use cases.

For example, lets say my monorepo looks like
```
> multiply
> subtract
> add
```
where `multiply` and `subtract` both depend on `add`. 

As a responsible developer, I add tests to all three of these workspaces. 

`add` is not compiled. In fact, none of these workspaces require a build step before testing. 

Now, I want all these tests to run in parallel. So, for each of these tasks I set `dependsOn: []`. After all, `multiply` tests do not depend on `add` tests. So these things are safe to run in parallel. 

Secondly, I want a change in `add` to force `multiply` tests to rerun. After all, a change in `add` might break the `multiply`. 

I don't think this is possible with the current configuration options! With `dependsOn: []`, code changes in `add` don't change the computed hash for `multiply`. With `dependsOn: ["^test"]`, my tests will no longer execute in parallel. 

To summarize, my requirements are
1. All tests run in parallel
2. A change in a workspace dependency results in a new hash for its dependents

### Expected Behavior

I expect two workspaces to be able to depend on each other from a code perspective (a change in one workspace affects the behavior of the other) without having the workspace tasks depend on each other (workspace A must finish this task before workspace B begins)

### To Reproduce

Follow the steps in the readme for https://github.com/rubber-duck-software/turborepo-caching-demo

### Reproduction Repo

https://github.com/rubber-duck-software/turborepo-caching-demo

**Comment by DerYeger** [comment]
I faced a similar issue with my monorepo.
Changes to my tsconfig and eslint packages did not invalidate the caches of tasks depending on either package.

As a workaround, I defined dummy build tasks for both packages:

```json
"@yeger/eslint-config#build": {
    "inputs": ["src/**"],
    "outputs": ["src/**"]
},
"@yeger/tsconfig#build": {
    "inputs": ["base.json", "lib.json", "web.json"],
    "outputs": ["base.json", "lib.json", "web.json"]
}
```

This approach does not require a dummy `build` script in the `package.json` files of both packages.

If you now add

```json
"add#build": {
  "inputs": ["src/**"],
  "outputs": ["src/**"],
},
"test": {
  "dependsOn": ["^build"]
}
```

to your pipeline, the `test` caches will invalidate once any of these files changes.
For your use case, you'd have to add the files of `add` to both its `inputs` and `outputs`.

**Comment by nathanhammond** [comment]
It's really clear what happens here if you take a look at how the graph gets built when it doesn't have a `dependsOn` relationship:
![no dependsOn](https://user-images.githubusercontent.com/20542/204821297-d3401305-e676-4f6f-b3e5-e4bbdd6e4434.png)

The issue here is that we don't have a way in our DSL to tell you to depend on your topological dependencies outside of a task. Or, in other words, `dependsOn: ["^"]`.

Fortunately, that's very straightforward to express as a dummy task in `turbo.json`:
```diff
 {
   "$schema": "https://turbo.build/schema.json",
   "pipeline": {
+    "topo": {
+      "dependsOn": ["^topo"]
+    },
     "test": {
+      "dependsOn": ["topo"]
     }
   }
 }
```

Once you do that you get the behavior you want and the graph looks like this:

![topo dependsOn](https://user-images.githubusercontent.com/20542/204823470-f7650df8-05de-4d5c-9b93-39feb2b58a74.png)

It's plausible that we should add "just ^" as an additional syntax rule in the DSL, but it's low priority for us to do that given that there is a clear method to accomplish this task (and the DSL would, under the hood, desugar to exactly this behavior).

***

<details>
```
~/repos/turborepo-caching-demo (main *)$ pnpm run test

> turborepo-caching-demo@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo
> turbo run test

• Packages in scope: add, multiply, subtract
• Running test in 3 packages
• Remote caching disabled
multiply:test: cache miss, executing a4071ac8184f70f3
add:test: cache miss, executing 2c815d3941833b6b
subtract:test: cache miss, executing a17e4fd5c8dad973
multiply:test: 
multiply:test: > multiply@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/multiply
multiply:test: > beartest
multiply:test: 
subtract:test: 
subtract:test: > subtract@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/subtract
subtract:test: > beartest
subtract:test: 
add:test: 
add:test: > add@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/add
add:test: > beartest
add:test: 
add:test: add 
subtract:test: subtract 
multiply:test: multiply 
multiply:test: ✓ 3 * 3 = 9
add:test: ✓ 1 + 1 = 2
subtract:test: ✓ 3 - 1 = 2
subtract:test: 
add:test: 
multiply:test: 

 Tasks:    3 successful, 3 total
Cached:    0 cached, 3 total
  Time:    10.342s 

~/repos/turborepo-caching-demo (main *)$ pnpm run test

> turborepo-caching-demo@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo
> turbo run test

• Packages in scope: add, multiply, subtract
• Running test in 3 packages
• Remote caching disabled
multiply:test: cache miss, executing d16ada726e2ef9d6
add:test: cache miss, executing 196aebf803d6ae02
subtract:test: cache miss, executing 44e59a442e41cf15
multiply:test: 
multiply:test: > multiply@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/multiply
multiply:test: > beartest
multiply:test: 
subtract:test: 
subtract:test: > subtract@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/subtract
subtract:test: > beartest
subtract:test: 
add:test: 
add:test: > add@1.0.0 test /Users/nathanhammond/repos/turborepo-caching-demo/add
add:test: > beartest
add:test: 
add:test: add 
multiply:test: multiply 
subtract:test: subtract 
multiply:test: 
multiply:test: ✗ 3 * 3 = 9 
multiply:test: 
subtract:test: 
subtract:test: ✗ 3 - 1 = 2 
subtract:test: 
add:test: 
add:test: ✗ 1 + 1 = 2 
add:test: 
add:test: AssertionError [ERR_ASSERTION]: 0 == 2
multiply:test: AssertionError [ERR_ASSERTION]: -9 == 9
multiply:test:     at /Users/nathanhammond/repos/turborepo-caching-demo/multiply/index.test.js:8:12
add:test:     at /Users/nathanhammond/repos/turborepo-caching-demo/add/index.test.js:8:12
add:test:     at async /Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:16:5
add:test:     at async RunSerially (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:5:5)
add:test:     at async runSuite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:33:5)
add:test:     at async suite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:71:5)
add:test:     at async Promise.all (index 0)
add:test:     at async runTests (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/testrunner.js:15:7) {
add:test:   generatedMessage: true,
add:test:   code: 'ERR_ASSERTION',
add:test:   actual: 0,
add:test:   expected: 2,
multiply:test:     at async /Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:16:5
multiply:test:     at async RunSerially (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:5:5)
multiply:test:     at async runSuite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:33:5)
multiply:test:     at async suite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:71:5)
multiply:test:     at async Promise.all (index 0)
multiply:test:     at async runTests (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/testrunner.js:15:7) {
multiply:test:   generatedMessage: true,
multiply:test:   code: 'ERR_ASSERTION',
multiply:test:   actual: -9,
multiply:test:   expected: 9,
multiply:test:   operator: '=='
multiply:test: }
add:test:   operator: '=='
add:test: }
subtract:test: AssertionError [ERR_ASSERTION]: 4 == 2
subtract:test:     at /Users/nathanhammond/repos/turborepo-caching-demo/subtract/index.test.js:8:12
subtract:test:     at async /Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:16:5
subtract:test:     at async RunSerially (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:5:5)
subtract:test:     at async runSuite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:33:5)
subtract:test:     at async suite (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/beartest.js:71:5)
subtract:test:     at async Promise.all (index 0)
subtract:test:     at async runTests (/Users/nathanhammond/repos/turborepo-caching-demo/node_modules/.pnpm/beartest-js@5.0.4/node_modules/beartest-js/testrunner.js:15:7) {
subtract:test:   generatedMessage: true,
subtract:test:   code: 'ERR_ASSERTION',
subtract:test:   actual: 4,
subtract:test:   expected: 2,
subtract:test:   operator: '=='
subtract:test: }
multiply:test:  ELIFECYCLE  Test failed. See above for more details.
subtract:test:  ELIFECYCLE  Test failed. See above for more details.
add:test:  ELIFECYCLE  Test failed. See above for more details.
add:test: ERROR: command finished with error: command (/Users/nathanhammond/repos/turborepo-caching-demo/add) pnpm run test exited (1)
subtract:test: ERROR: command finished with error: command (/Users/nathanhammond/repos/turborepo-caching-demo/subtract) pnpm run test exited (1)
command (/Users/nathanhammond/repos/turborepo-caching-demo/add) pnpm run test exited (1)
command (/Users/nathanhammond/repos/turborepo-caching-demo/subtract) pnpm run test exited (1)

 Tasks:    0 successful, 2 total
Cached:    0 cached, 2 total
  Time:    10.304s 

 ERROR  run failed: command  exited (1)
 ELIFECYCLE  Test failed. See above for more details.
```
</details>

**Comment by nathanhammond** [comment]
Conceptually `dependsOn` actually conflates two things: package dependencies (`^`, `docs#`) and task dependencies (`build`). Both need to be specified to get the behavior a person may want.

The reason for the distinction here is that _some_ tasks (think: apps that have bundlers that will do the right thing for dependencies throughout the application, or lint tasks that are simply static analysis of all code unaware of the project architecture) don't need to be aware of any topological dependencies. We can already express those in our DSL gracefully.

Your example is the inverse: topological dependencies, but no task dependencies. We can make our DSL do this, but it requires the `^topo` indirection workaround.

**Comment by rubber-duck-software** [comment]
Ahh- that example is very helpful. A tool like eslint will not be affected by changes in a dependency, but a tool like tsc will. 

As for the `dependsOn` option, if I understand you correctly, you are saying it expresses two things.
1. This task is dependent on the following tasks in the same workspace (`build`)
2. This task is dependent on the following tasks in other workspaces (`^build`, `docs#`)

If I understand the current configuration correctly from the docs

**dependsOn**: The list of tasks this task depends on.
**env**: The list of environment variables this task depends on.
**inputs**: The list of files this task depends on. (defaults to all files in the workspace)

If I understand the `topo` workaround correctly, it works because `multiply#test` depends on `add#topo` which in turn depends on all the files in `add` workspace via the `inputs` option.

An equivalent way of solving this (though far more fragile) would be 

```json
"test": {
    "dependsOn": []
},
"multiply#test": {
    "inputs": ["../add/*"]
},
"subtract#test": {
    "inputs": ["../add/*"]
}
```

Now understanding that, I guess I would phrase the request as

Add the ability to express file dependencies (like the `input` option) but using pnpm filter syntax for packages (like the `dependsOn` option).

**Comment by nathanhammond** [comment]
Your alternative proposal is _not quite_ equivalent. You're still not taking advantage of the dependency graph. You're instead specifying a list of inputs which can be individually considered as part of a cache key for a task.

In other words, it's still this:
![Screen Shot 2022-11-30 at 22 24 56](https://user-images.githubusercontent.com/20542/205223870-fb362804-e733-47df-a2aa-57dd974a29d1.png)

The trick in the `^topo` approach is that you can _also_ specify `"add#topo": { "inputs": ["src/**"] }` and then all things in the `topo` graph will do the right thing. (This is a thing I had on my mind for why _not_ ship the "just `^`" solution.)

From a Turborepo perspective the inputs pattern you specify has one additional major drawback: each one of those input globs must be individually calculated. Putting it into a tree means it's calculated just once.

We're _definitely_ not going to add a way to express dependencies that are in a separate package like this, nor provide a way to make that less painful by using filter syntax. We want to encourage people into thinking in graphs.

**Comment by rubber-duck-software** [comment]
Good point- I had not considered the performance impacts of my proposal. 

So, what I think I'm hearing is 
1. We probably do not want to introduce new syntax to handle this use case
2. We definitely do not want to introduce a new configuration option to handle this use case. 

If that is true, it would seem the "workaround" is no longer a workaround but instead the long term way of solving this. Thanks for the PR around documenting this @anthonyshew. 

I am also going to create a PR for adding documentation to the Monorepo Handbook under linting. I think running `tsc` across all workspaces will require this workaround.

**Comment by nathanhammond** [comment]
I am still open to being convinced that "just `^`" is a worthwhile feature which is why I tweaked this issue instead of closed it.

It's not new syntax, it's not a weird overload, and being able to explain everything as a cross product of options is also useful pedagogically.

But I do want to give it time and see precisely how many people head down this path, and see if other features we end up building start to work with it in a graceful way.

We pay a maintenance tax on every public API we support, so I try to be conservative about adding it.

**Comment by rubber-duck-software** [comment]
As for the number of people who head down this path, I am going to guess this is common because of typescript.

The monorepo handbook discusses using typescript as a linter. I'd say any monorepo which uses typescript as a linter and wants to run typescript in parallel across their packages will run into this. 

See example repo https://github.com/rubber-duck-software/turborepo-tsc-caching


Regarding the "^", I think it's a bit too subtle for my comfort, but I'm only one user. It could very well be introducing "^" is the best answer. This would clean up the most common examples of this issue at any rate and the `topo` workaround could be used for the remainder.

**Comment by DerYeger** [comment]
> As for the number of people who head down this path, I am going to guess this is common because of typescript.
> 
> The monorepo handbook discusses using typescript as a linter. I'd say any monorepo which uses typescript as a linter and wants to run typescript in parallel across their packages will run into this.
> 
> See example repo https://github.com/rubber-duck-software/turborepo-tsc-caching
> 
> Regarding the "^", I think it's a bit too subtle for my comfort, but I'm only one user. It could very well be introducing "^" is the best answer. This would clean up the most common examples of this issue at any rate and the `topo` workaround could be used for the remainder.

From your example:

>Observed that all tsc tasks succeed because they use a cached result. Desired behavior is that we would not get a cached result but instead all tests would fail. 

Isn't this issue expected? Your pipeline is deliberately running `tsc` in parallel. Considering one package depends on the other, a task dependency is, IMO, required.

That's also how I organized my [monorepo](https://github.com/DerYeger/yeger):

```json
"typecheck": {
    "dependsOn": ["^build"],
    "inputs": [
      "index.html",
      "tsconfig.json",
      "vite.config.ts",
      "*.d.ts",
      "public/**",
      "src/**",
      "test/**"
    ],
    "outputs": []
  }
```

with `typecheck` running `tsc --noEmit`.

**Comment by rubber-duck-software** [comment]
It is possible to run tsc in parallel and still get accurate type check failures. Per @nathanhammond's suggestions above, the correct turbo.json file for this would have been 

```json
{
  "$schema": "https://turborepo.org/schema.json",
  "pipeline": {
    "topo": {
      "dependsOn": ["^topo"]
    },
    "tsc": {
      "dependsOn": ["^topo"]
    }
  }
}
```

You can see this example working at https://github.com/rubber-duck-software/turborepo-tsc-caching/tree/working-example

This configuration will both run tsc in parallel across all projects and will ensure correct caching behavior. However, the configuration is subtle enough that I think most users will miss it and accidentally configure only partial parallelization or (worse) incorrect caching.

**Comment by DerYeger** [comment]
> It is possible to run tsc in parallel and still get accurate type check failures. Per @nathanhammond's suggestions above, the correct turbo.json file for this would have been
> 
> ```json
> {
>   "$schema": "https://turborepo.org/schema.json",
>   "pipeline": {
>     "topo": {
>       "dependsOn": ["^topo"]
>     },
>     "tsc": {
>       "dependsOn": ["^topo"]
>     }
>   }
> }
> ```
> 
> You can see this example working at https://github.com/rubber-duck-software/turborepo-tsc-caching/tree/working-example
> 
> This configuration will both run tsc in parallel across all projects and will ensure correct caching behavior. However, the configuration is subtle enough that I think most users will miss it and accidentally configure only partial parallelization or (worse) incorrect caching.

Yes, across all projects, but not dependency and dependent.
If you change it to

```tsc
"tsc": {
  "dependsOn": ["^tsc"]
}
```

it will still run in parallel for unrelated projects (i.e., no dependency on each other).
The result will be identical to your `topo` approach.

**Comment by rubber-duck-software** [comment]
Hi @DerYeger! Thanks for continuing to engage on this. 


Your example of 
```json
"tsc": {
  "dependsOn": ["^tsc"]
}
```

is different than the `topo` approach. Let's say in my example repo tsc takes 10 seconds for `add` and 10 seconds for `subtract`. If turbo.json were authored as you suggest, total time to type check the monorepo would be ~20 seconds. Using the `topo` approach the total time would be ~10 seconds. Both approaches ensure correct caching. 

I have added a some sleep time into https://github.com/rubber-duck-software/turborepo-tsc-caching/tree/working-example so you can verify this behavior for yourself. 

My opinion is that this is a lot of subtlety for a fairly vanilla example. I am happy to help with documentation on this. :)

**Comment by DerYeger** [comment]
Ah, my bad. Because of `tsc`, I immediately jumped to assumption that your `subtract` depends on the `dist` files of `add`.

**Comment by anthonyshew** [comment]
@rubber-duck-software @DerYeger [Docs](https://turbo.build/repo/docs/core-concepts/monorepos/running-tasks#dependencies-outside-of-a-task) for this behavior have landed!

The opinion I've developed working through this issue and doccing is that both options (`"^"` or the dummy task) feel like they bring equal clarity to a reader of a `turbo.json` file. If I were stumbling upon either of these in a `turbo.json` and had never seen them before, I would have to learn what either of them really mean. `"^"` feels something like a "magic string" to me but, if I know how Turbo tasks work with enough depth, I _may_ be able to grok what the dummy task strategy does - and, if I can't, now it's docced.

@nathanhammond @tknickman Will leave it to y'all's judgment as far as if we think a DSL change is in order or if we can close this. 👍

**Comment by rubber-duck-software** [comment]
I believe #2911 is still relevant. Any feedback would be appreciated.

**Comment by connorjs** [comment]
@nathanhammond - In https://github.com/vercel/turbo/issues/2855#issuecomment-1334793695 you mentioned

> We want to encourage people into thinking in graphs.

I did not understand this `topo` “trick” the first time I read it in the docs, but I understand it more now from this issue. As I went to add it to my repo, I considered the alternate name `dependencies` (for the nice `"dependsOn": ["dependencies"]`).

Any comments on the word choice here? Does `topological` “speak this idea better/more precisely” than `dependencies`?

**Comment by nathanhammond** [comment]
@connorjs You could call it "get-your-flu-shot" and it'll still work; `dependencies` has a naming collision with `package.json` that can be confusing for people so I chose `topo` to make it clear that it was distinct.
